### DISCUSSION TITLE
Installing additional RPM packages removes keycloak user

### DISCUSSION THREAD

**Discussion by MihaStemberger** [keycloak/keycloak#23166] [Q&A] [answered]
Hello!

I am in need to install `jq`(https://jqlang.github.io/jq/) into Keycloak docker container. While following https://www.keycloak.org/server/containers#_installing_additional_rpm_packages instructions, I was unable to do so without altering the code. Error that started to show up was: `unable to find user keycloak: no matching entries in passwd file`.

`Keycloak version: 21.1.2`
`Docker version: 24.0.4, build 3713ee1`

**Original code** from instructions(added wanted versions and entypoint):
```dockerfile
FROM registry.access.redhat.com/ubi9:9.2 AS ubi-micro-build
RUN mkdir -p /mnt/rootfs
RUN dnf install --installroot /mnt/rootfs jq --releasever 9 --setopt install_weak_deps=false --nodocs -y;  \
    dnf --installroot /mnt/rootfs clean all;

FROM quay.io/keycloak/keycloak:21.1.2
COPY --from=ubi-micro-build /mnt/rootfs /

ENTRYPOINT ["/opt/keycloak/bin/kc.sh", "start-dev"]
```
The above code ends up with no keycloak user left inside the container, here are some observations:

1. No keycloak in `/etc/group`
```text
bash-5.1$ cat /etc/group 
root:x:0:
bin:x:1:
daemon:x:2:
sys:x:3:
adm:x:4:
tty:x:5:
disk:x:6:
lp:x:7:
mem:x:8:
kmem:x:9:
wheel:x:10:
cdrom:x:11:
mail:x:12:
man:x:15:
dialout:x:18:
floppy:x:19:
games:x:20:
tape:x:33:
video:x:39:
ftp:x:50:
lock:x:54:
audio:x:63:
users:x:100:
nobody:x:65534:
```
2. No keycloak in `/etc/passwd`
```text
bash-5.1$ cat /etc/passwd 
root:x:0:0:root:/root:/bin/bash
bin:x:1:1:bin:/bin:/sbin/nologin
daemon:x:2:2:daemon:/sbin:/sbin/nologin
adm:x:3:4:adm:/var/adm:/sbin/nologin
lp:x:4:7:lp:/var/spool/lpd:/sbin/nologin
sync:x:5:0:sync:/sbin:/bin/sync
shutdown:x:6:0:shutdown:/sbin:/sbin/shutdown
halt:x:7:0:halt:/sbin:/sbin/halt
mail:x:8:12:mail:/var/spool/mail:/sbin/nologin
operator:x:11:0:operator:/root:/sbin/nologin
games:x:12:100:games:/usr/games:/sbin/nologin
ftp:x:14:50:FTP User:/var/ftp:/sbin/nologin
nobody:x:65534:65534:Kernel Overflow User:/:/sbin/nologin
```
3. `whoami`
```text
bash-5.1$ whoami
whoami: cannot find name for user ID 1000
``` 

**Altered code**(added `rm -rf /mnt/rootfs/etc/group /mnt/rootfs/etc/passwd`):
```dockerfile
FROM registry.access.redhat.com/ubi9:9.2 AS ubi-micro-build
RUN mkdir -p /mnt/rootfs
RUN dnf install --installroot /mnt/rootfs jq --releasever 9 --setopt install_weak_deps=false --nodocs -y;  \
    dnf --installroot /mnt/rootfs clean all; \
    rm -rf /mnt/rootfs/etc/group /mnt/rootfs/etc/passwd;

FROM quay.io/keycloak/keycloak:21.1.2
COPY --from=ubi-micro-build /mnt/rootfs /

ENTRYPOINT ["/opt/keycloak/bin/kc.sh", "start-dev"]
```
With this code keycloak user and group still exist.

My question, is a valid and OK way to install `jq`, or am I missing something?

Thank you and best regards,
Miha

**Comment by vmuzikar** [comment]
@ASzc Could this be a bug maybe?

**Reply by ASzc** [reply]
Possibly, but I would have thought we'd have noticed that from the beginning. I can have a look at it

**Reply by ASzc** [reply]
I don't get the same result with `podman`, i.e. it works without the 1000 user being missing. Possibly this is one of those things that `docker` handles poorly. I suspect kubernetes/OpenShift wouldn't have a problem either, as it does user replacement similar to podman.

I do note that in my case the username changed however:

```
aszczucz@tyrfing ~/s/k/k/q/container (main)> podman run --rm -ti --entrypoint whoami keycloak:999-SNAPSHOT
keycloak
aszczucz@tyrfing ~/s/k/k/q/container (main)> podman run --rm -ti --entrypoint whoami keycloak-custom:999-SNAPSHOT
aszczucz
```

So there is some truth here, I just can't reproduce it. A slightly cleaner way to get rid of the offending files might be adding:

```
rpm -e --nodeps setup
```

To the first build stage, after dnf

**Reply by SomeDeveloper13** [reply]
I'm also using docker, not podman.
I tried using the `rpm` command in a `RUN`, but it didn't recognize the executable. (Most likely needs an extra step, or I put it in the wrong place)
Using the `rm` command as @MihaStemberger suggested worked for me, though.

Might be a good idea to update the instructions as this caused me quite the headache till I found this thread.
https://www.keycloak.org/server/containers

**Reply by ASzc** [reply, marked as answer]
Ok, I've filed a bug for it (#23637). Will need to have someone from here test the MR as I'm unable to reproduce.

**Reply by SomeDeveloper13** [reply]
Left a comment on the PR that the updates to the instructions restored the `keycloak` user and the $HOME env variable.
https://github.com/keycloak/keycloak/pull/23639#issuecomment-1743427806
